### PR TITLE
Completion Support

### PR DESCRIPTION
**Pull request by TylerLeonhardt**

### CODE REVIEW

File: src/PowerShellEditorServices.Engine/Services/TextDocument/CompletionResults.cs [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+                case CompletionResultType.Variable:
+                    return CompletionType.Variable;
+
+                case CompletionResultType.Namespace:
+                    return CompletionType.Namespace;
+
+                case CompletionResultType.Type:
+                    return CompletionType.Type;
+
+                case CompletionResultType.Keyword:
```
Comment (rkeithhill) [inline src/PowerShellEditorServices.Engine/Services/TextDocument/CompletionResults.cs]: I wonder if we should map `DynamicKeyword` to `CompletionType.Keyword`?

File: test/Pester/EditorServices.Integration.Tests.ps1 [hunk 2]
```diff
@@ -464,6 +464,22 @@ Get-Foo
         $edit.Arguments.EndColumnNumber | Should -Be 4
     }
 
+    It "handles textDocument/completion properly" {
```
Comment (rkeithhill) [inline test/Pester/EditorServices.Integration.Tests.ps1]: There's plenty of opportunities here to add more completion tests.  :-)  Thinking of parameter names, paths with spaces, variable names, etc.

File: src/PowerShellEditorServices.Engine/Services/TextDocument/CompletionResults.cs (hunk 1 above)
Comment (TylerLeonhardt) [reply src/PowerShellEditorServices.Engine/Services/TextDocument/CompletionResults.cs]: Do you think those are falling into `CompletionType.Unknown`?

File: test/Pester/EditorServices.Integration.Tests.ps1 (hunk 2 above)
Comment (TylerLeonhardt) [reply test/Pester/EditorServices.Integration.Tests.ps1]: For sure! These e2e tests will act as smoke tests. After everything is ported over, I add tests for every handler as unit tests.

File: src/PowerShellEditorServices.Engine/Services/TextDocument/CompletionResults.cs (hunk 1 above)
Comment (rkeithhill) [reply src/PowerShellEditorServices.Engine/Services/TextDocument/CompletionResults.cs]: I would suppose so but to be honest this is a feature (dynamic keywords) where I have little experience.  Just seemed like low hanging fruit. :-)

Comment (TylerLeonhardt) [reply src/PowerShellEditorServices.Engine/Services/TextDocument/CompletionResults.cs]: I wonder if that could be causing some issues with loading of completions of dynamic parameters... probably not. I'll still add it. Thanks!

**Comment by rkeithhill** [comment]
Will this only support v5 AST and later?  If so, can we finally add class/enum icons for completion intellisense?  Also, will want to add that for document symbols.  Perhaps in a separate PR though.

**Comment by TylerLeonhardt** [comment]
> Will this only support v5 AST and later?

Yep!

> If so, can we finally add class/enum icons for completion intellisense?

This _should_ work today because the API hasn't changed... I think.

Where we're lacking on class/enum support is symbols used in references. I could be wrong here.

**Comment by TylerLeonhardt** [comment]
rebased on the new C# E2E tests and things look good!

File: src/PowerShellEditorServices.Engine/Services/TextDocument/CompletionResults.cs (hunk 1 above)
Comment (TylerLeonhardt) [reply src/PowerShellEditorServices.Engine/Services/TextDocument/CompletionResults.cs]: added!

Comment (rjmholt) [reply src/PowerShellEditorServices.Engine/Services/TextDocument/CompletionResults.cs]: Probably not causing dynamic parameter issues. Unless you're using DSC, it's unlikely you'd see any issue here

**Pull request merged**